woocommerce/migrator-cli
# Add command to save PAN import data

This PR adds a command to update payment methods from shopify_payments to WooPayments using a mapping file generated by stripe

**Test instructions**

If this server PR has not been merged to trunk yet make sure to checkout to that branch first
https://github.com/Automattic/woocommerce-payments-server/pull/4159

- Checkout to this PR's branch

**Creating a stripe customer and card:**
- Create a stripe customer with a card attached to it by placing an order and checking the option to save the card during checkout. Use the card `4242 4242 4242 4242`
- Take note of the token `pm_`  from `wp_woocommerce_payment_tokens` for this user
- Remove the entries for this user from `wp_woocommerce_payment_tokens` table
- Take note of the `wp__wcpay_customer_id_test` from `wp_usermeta` for this user
- Remove the `wp__wcpay_customer_id_test` entry from `wp_usermeta` table


**Mocking an Shopify Payments order**
Execute this steps if you don't have a Shopify Payments order and skip the **Importing an order** steps
- Create an order with a subscription product for the user created in the **Creating a stripe customer and card:** steps
- If a subscription was created delete it but keep the order
- Add `_order_number` to the `wp_post_meta` table and set the value to `11664` for this order
- Update the order creation date via the /wp-admin order edit page to `2023-01-07`
- Add `_original_payment_last_4` to the `post_meta` and set the value to `4242`
- Add `_original_payment_method_id` to the `post_meta` and set the value to `pm_mock`
- Add `_original_payment_gateway` to the `post_meta` and set the value to `shopify_payments`

**Importing an order**
- You won't be able to test the order import part if you don't have a Shopify Payments order.
- Take note of the customer email you created in the **Creating a stripe customer and card** steps
- Change the customer email to something else via wp-admin
- Execute: `wp migrator orders --ids="<shopify_order_id>`
- Make sure to change the `shopify_order_id` to an order that was placed using shopify payments
- Make sure to enable subscriptions for the product in the order in case it's not enabled yet.
- Change the `_order_number` in the `wp_post_meta` table to `11664` for this order
- Update the order creation date via the /wp-admin order edit page to `2023-01-07`
- Change the value of `_original_payment_last_4` in the `post_meta` table to `4242`
- Take note of the `_original_payment_method_id` from `post_meta` table
- Change the order user email to the one you took note in the previous steps.

**Importing payment methods**
- Execute the following command `wp migrator import_stripe_data_into_woopayments ` to import the customer data from stripe
- Make sure the new token was created in `wp_woocommerce_payment_tokens` for the order user

**Importing a Subscription **
- Download these files
[orders.json](https://github.com/woocommerce/migrator-cli/files/13429838/orders.json)
[subscriptions.json](https://github.com/woocommerce/migrator-cli/files/13430581/subscriptions.json)

- Run `wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path>`
- Make sure the `_original_payment_gateway`, `_original_payment_method_id` and `_original_payment_last_4` in the `post_meta` table matches the ones in the order imported/created before

**Updating the payment methods in the subscription and order**
- Download the following file:
[migration.csv](https://github.com/woocommerce/migrator-cli/files/13430002/migration.csv)
Change the `<source_id_old>` to match the `_original_payment_method_id`, the `<customer_id_new>` to match the stripe customer id `cus_` you took note earlier and the `<payment_method_new>` to match the stripe payment method id `pm_` you took note earlier
- Execute the following command: 
`wp migrator add_woopayments_migration_data --migration_file=<migration_file_abs_path>`
- Now make sure `Payment Method:` is set to `Credit card / Debit card` and `Saved payment method:` is set to the correct card  in the billing data for the subscription
- Trigger a subscription renewal via the subscription actions
- Make sure the new order was charged correctly
